EIPS/eip-6785.md

```diff
@@ -0,0 +1,203 @@
+---
+eip: 6785
+title: ERC-721 Utilities Extension
+description: Easy access to utility information via functions and the NFT's metadata
+author: Otniel Nicola (@OT-kthd), Bogdan Popa (@BogdanKTHD)
+discussions-to: https://ethereum-magicians.org/t/eip-6785-erc-721-utilities-extension/13568
+status: Draft
+type: Standards Track
+category: ERC
+created: 2023-03-27
+requires: 165, 721
+---
+
+## Abstract
+
+This specification defines standard functions that outline what a utility entails for a specific NFT and how it may be
+used. This specification is an optional extension of [ERC-721](./eip-721.md).
+
+## Motivation
+
+This specification aims to bring clarity what the utility associated with an NFT is and how to access this utility.
+Relying on third-party platforms to obtain information regarding the utility of the NFT that one owns can lead to scams,
+phishing or other forms of fraud.
+
+## Specification
+
+The keywords “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and
+“OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+Every contract compliant with [ERC-6785](./eip-6785.md) MUST implement the interface defined as follows:
+
+### Contract Interface
+
+```solidity
+// @title NFT Utility description
+///  Note: the EIP-165 identifier for this interface is 0xbcc08bd0
+
+interface IERC6785 {
+
+    // Logged when the utility description URL of an NFT is changed
+    /// @notice Emitted when the utilityURL of an NFT is changed
+    /// The empty string for `utilityUri` indicates that there is no utility associated
+    event UpdateUtility(uint256 indexed tokenId, string utilityUri);
+
+    /// @notice set the new utilityUri - remember the date it was set on
+    /// @dev The empty string indicates there is no utility
+    /// Throws if `tokenId` is not valid NFT
+    /// @param utilityUri  The new utility description of the NFT
+    /// 324a28b4
+    function setUtilityUri(uint256 tokenId, string utilityUri) external;
+
+    /// @notice Get the utilityUri of an NFT
+    /// @dev The empty string for `utilityUri` indicates that there is no utility associated
+    /// @param tokenId The NFT to get the user address for
+    /// @return The utility uri for this NFT
+    /// 438b8582
+    function getUtilityUri(uint256 tokenId) external view returns (string memory);
+
+    /// @notice Get the changes made to utilityUri
+    /// @param tokenId The NFT to get the user address for
+    /// @return The history of changes to `utilityUri` for this NFT
+    /// af5a3ae2
+    function getUtilityHistory(uint256 tokenId) external view returns (string[] memory);
+}
+```
+
+All functions defined as view MAY be implemented as pure or view
+
+Function `setUtilityUri` MAY be implemented as public or external
+
+The event `UpdateUtility` MUST be emitted when the setUtilityUri function is called
```

Comment:
Should it also be emitted for changes not caused by `setUtilityUri`? Say in the constructor, or if a contract has some custom batch update function?